torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -0,0 +1,191 @@
+            }
+          }
+          ++weightOffset;
+        }
+      }
+    }
+    gradInput.data()[linearIndex] = value;
+  }
+}
+
+template <typename T, typename IndexType>
+__global__ void spatialDepthwiseConvolutionAccGradParameters(
+    const THCDeviceTensor<T, 4> gradOutput,
+    const THCDeviceTensor<T, 4> input,
+    THCDeviceTensor<T, 4> gradWeight,
+    const int batchSize,
+    const int inputChannels,
+    const int kernelChannels,
+    const int depthwiseMultiplier,
+    IndexType blockElements,
+    const int inputWidth, const int inputHeight,
+    const int outputWidth, const int outputHeight,
+    const int kernelWidth, const int kernelHeight,
+    const int strideWidth, const int strideHeight,
+    const int padWidth, const int padHeight,
+    const int dilationWidth, const int dilationHeight)
+{
+  // Have to use a statically typed Shared Memory pointer
+  SharedMem<T> smem;
+
+  // Each Block is responsible for accumulating over a permutation of
+  // (channels x kH x kW), use blockIdx to determine which one
+  int bidx = blockIdx.x;
+  int kW = bidx % kernelWidth;
+  int kH = (bidx / kernelWidth) % kernelHeight;
+  int ch = (bidx / kernelWidth / kernelHeight) % kernelChannels;
+
+  // Need to calculate which input channel is associated with this filter
+  // channel
+  int inputCh = ch / depthwiseMultiplier;
+
+  T grad = ScalarConvert<float, T>::to(0.0);
+
+  // Block-stride loop over the number of elements we need to reduce
+  for (IndexType idx = threadIdx.x; idx < blockElements; idx += blockDim.x) {
+    // Need to calculate the following: batch position, and offset into the gradOutput
+    // in height, and width. We can intuit the corresponding position in the input from
+    // the other parameters we have
+    int go_w_offset = idx % outputWidth;
+    int go_h_offset = (idx / outputWidth) % outputHeight;
+    int batch = (idx / outputWidth / outputHeight) % batchSize;
+
+    int i_w_offset = (go_w_offset * strideWidth) + (kW * dilationWidth) - padWidth;
+    int i_h_offset = (go_h_offset * strideHeight) + (kH * dilationHeight) - padHeight;
+
+    if (i_w_offset >= 0 && i_h_offset >= 0 && i_w_offset < inputWidth && i_h_offset < inputHeight) {
+      int inputOffset = ((batch * inputChannels + inputCh) * inputHeight + i_h_offset) * inputWidth + i_w_offset;
+      int outputOffset = ((batch * kernelChannels + ch) * outputHeight + go_h_offset) * outputWidth + go_w_offset;
+      grad = THCNumerics<T>::add(
+          grad,
+          THCNumerics<T>::mul( input.data()[inputOffset], gradOutput.data()[outputOffset]));
+    }
+  }
+  __syncthreads();
+
+  // At this point each thread in the block has a local gradient, which we need to
+  // accumulate prior to writing the global value
+  T *buf = smem.getPointer();
+  T tval = reduceBlock<T, ReduceAdd<T, T>>(
+      buf, blockDim.x, grad, ReduceAdd<T, T>(), ScalarConvert<float, T>::to(0));
+
```

Comment:
reduction through shared memory (in reduceBlock) is usually slower than through warp intrinsics, but it's likely not a bottleneck here.

Reply:
I have had an open PR for using a warp shuffle in CuTorch reductions but it never got merged: https://github.com/torch/cutorch/pull/750

Reply:
We could also start using CUB for that